crates/ra_assists/src/assists/early_return.rs

```diff
@@ -0,0 +1,275 @@
+//! FIXME: write short doc here
+
+use crate::{
+    assist_ctx::{Assist, AssistCtx},
+    AssistId,
+};
+use hir::db::HirDatabase;
+use itertools::Itertools;
+use ra_syntax::{
+    ast::Block,
+    ast::ContinueExpr,
+    ast::FnDef,
+    ast::IfExpr,
+    ast::LoopExpr,
+    ast::ReturnExpr,
+    ast::WhileExpr,
+    AstNode,
+    SyntaxKind::{L_CURLY, R_CURLY, WHITESPACE},
+    SyntaxToken,
+};
+
+pub(crate) fn convert_to_guarded_return(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let if_expr: IfExpr = ctx.node_at_offset()?;
+    let expr = if_expr.condition()?.expr()?;
+    let then_block = if_expr.then_branch()?.block()?;
+    if if_expr.else_branch().is_some() {
+        return None;
+    }
+
+    let parent_block = if_expr.syntax().parent()?.ancestors().find_map(Block::cast)?;
+
+    if parent_block.expr()? != if_expr.clone().into() {
+        return None;
+    }
+
+    // check for early return and continue
+    let first_in_then_block = then_block.syntax().first_child()?.clone();
+    if ReturnExpr::can_cast(first_in_then_block.kind())
+        || ContinueExpr::can_cast(first_in_then_block.kind())
+        || first_in_then_block
+            .children()
+            .any(|x| ReturnExpr::can_cast(x.kind()) || ContinueExpr::can_cast(x.kind()))
+    {
+        return None;
+    }
+
+    let parent_container = parent_block.syntax().parent()?.parent()?;
+    let early_expression = if LoopExpr::can_cast(parent_container.kind())
+        || WhileExpr::can_cast(parent_container.kind())
+    {
+        Some("continue;")
+    } else if FnDef::can_cast(parent_container.kind()) {
+        Some("return;")
+    } else {
+        None
+    }?;
+
+    if then_block.syntax().first_child_or_token().map(|t| t.kind() == L_CURLY).is_none() {
+        return None;
+    }
+
+    let last = then_block.syntax().last_child_or_token().filter(|t| t.kind() == R_CURLY)?;
+    let cursor_position = ctx.frange.range.start();
+
+    ctx.add_action(AssistId("convert_to_guarded_return"), "convert to guarded return", |edit| {
+        let statements = then_block
```

Comment:
I tried to copy children_with_tokens(), but it wasn't clear how to remove update indents. So I decided to try this one for now, but maybe I need to import rowan and use match in the map to make it cleare.

Reply:
Yeah, what we ideally need here is to walk the tree and to modify each leading whitespace token. Ie, something like this:

```
fn deindent(el: &SyntaxNode) -> SyntaxNode {
    // walk the tree, collect all whitespace nodes with `\n`, call `algo::replace_children`
}
```